snorkel/slicing/modules/slice_combiner.py

```diff
@@ -44,83 +44,79 @@ def __init__(
-        # into a [batch_size x num_slices] tensor
         indicator_preds = torch.cat(
             [
-                F.softmax(outputs[slice_ind_name][0])[:, 0].unsqueeze(1)
-                for slice_ind_name in slice_ind_op_names
+                F.softmax(output, dim=1)[:, 0].unsqueeze(1)
+                for output in indicator_outputs
             ],
             dim=-1,
         )
-        predictor_preds = torch.cat(
+
+        # Concatenate predictor head confidences into tensor [batch_size x num_slices]
+        predictor_outputs = self._collect_flow_outputs_by_key(
+            flow_dict, self.slice_pred_key
+        )
+
+        predictor_confidences = torch.cat(
             [
-                F.softmax(outputs[slice_pred_name][0])[:, 0].unsqueeze(1)
-                for slice_pred_name in slice_pred_op_names
+                # Compute the "confidence" using the max score across classes
+                torch.max(F.softmax(output, dim=1), dim=1)[0].unsqueeze(1)
+                for output in predictor_outputs
             ],
             dim=-1,
         )
 
-        # Collect names of predictor "features" that will be combined into the final
-        # reweighted representation
-        slice_feat_names = sorted(
-            [
-                flow_name
-                for flow_name in outputs.keys()
-                if self.slice_pred_feat_key in flow_name
-            ]
+        # Concatenate each predictor feature (to be combined into reweighted
+        # representation) into [batch_size x 1 x feat_dim] tensor
+        predictor_feat_outputs = self._collect_flow_outputs_by_key(
+            flow_dict, self.slice_pred_feat_key
         )
-
-        # Concatenate each predictor feature into [batch_size x 1 x feat_dim] tensor
         slice_representations = torch.cat(
-            [
-                outputs[slice_feat_name][0].unsqueeze(1)
-                for slice_feat_name in slice_feat_names
-            ],
-            dim=1,
+            [output.unsqueeze(1) for output in predictor_feat_outputs], dim=1
         )
 
         # Attention weights used to combine each of the slice_representations
         # incorporates the indicator (whether we are in the slice or not) and
         # predictor (confidence of a learned slice head)
-        A = F.softmax(indicator_preds * predictor_preds, dim=1)
+        A = F.softmax(indicator_preds * predictor_confidences, dim=1)
 
-        # Match the dimensions of the slice_representations
+        # Match dims [bach_size x num_slices x feat_dim] of slice_representations
         A = A.unsqueeze(-1).expand([-1, -1, slice_representations.size(-1)])
 
-        # Reweight representations by class Sum across all classes
+        # Reweight representations with weighted sum across slices
         reweighted_rep = torch.sum(A * slice_representations, dim=1)
         return reweighted_rep
+
+    def _collect_flow_outputs_by_key(
```

Comment:
Should this be module level? No self attributes accessed

Reply:
alternative is as a more general helper function in `snorkel/classifier/utils`?

Reply:
done^